File: backend/app/routers/utang.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from app.database import get_db
from app.models.utang import Utang
from app.utils import get_wita_now

router = APIRouter(prefix="/api/utang", tags=["Utang"])

class UtangSchema(BaseModel):
    nama_pelanggan: str = Field(min_length=1, max_length=200)
    jumlah: float = Field(ge=0)
    keterangan: Optional[str] = Field(default=None, max_length=500)

class UtangResponse(BaseModel):
    id: int
    nama_pelanggan: str
    jumlah: float
    keterangan: Optional[str]
    tanggal: Optional[datetime]
    lunas: bool
    tanggal_lunas: Optional[datetime]

    class Config:
        from_attributes = True

@router.get("/")
def get_utang(status: str = Query("semua", pattern="^(semua|belum|lunas)$"), db: Session = Depends(get_db)):
    query = db.query(Utang)
    if status == "belum":
        query = query.filter(Utang.lunas == False)
    elif status == "lunas":
        query = query.filter(Utang.lunas == True)
    return query.order_by(Utang.tanggal.desc()).all()

@router.get("/ringkasan")
def get_ringkasan_utang(db: Session = Depends(get_db)):
    belum_lunas = db.query(Utang).filter(Utang.lunas == False).all()
    total = sum(u.jumlah for u in belum_lunas)
    return {
        "total_utang": total,
        "jumlah_pelanggan": len(belum_lunas),
    }

@router.post("/")
def tambah_utang(utang: UtangSchema, db: Session = Depends(get_db)):
    db_utang = Utang(**utang.model_dump())
    db.add(db_utang)
    db.commit()
    db.refresh(db_utang)
    return db_utang

@router.put("/{id}/lunas")
def tandai_lunas(id: int, db: Session = Depends(get_db)):
    db_utang = db.query(Utang).filter(Utang.id == id).first()
    if not db_utang:
        raise HTTPException(status_code=404, detail="Data utang tidak ditemukan")
    
    db_utang.lunas = True
    db_utang.tanggal_lunas = get_wita_now()
    db.commit()
    db.refresh(db_utang)
    return db_utang

File: frontend/src/stores/utang.ts
import { defineStore } from "pinia";
import api from "@/api";

interface UtangItem {
  id: number;
  nama_pelanggan: string;
  jumlah: number;
  keterangan: string | null;
  tanggal: string;
  lunas: boolean;
  tanggal_lunas: string | null;
}

interface UtangRingkasan {
  total_utang: number;
  jumlah_pelanggan: number;
}

export const useUtangStore = defineStore("utang", {
  state: () => ({
    utangList: [] as UtangItem[],
    ringkasan: { total_utang: 0, jumlah_pelanggan: 0 } as UtangRingkasan,
    loading: false,
    error: null as string | null,
  }),

  actions: {
    async fetchUtang(status: string = "semua") {
      this.loading = true;
      this.error = null;
      try {
        const res = await api.get(`/api/utang/?status=${status}`);
        this.utangList = res.data;
      } catch (e) {
        this.error = "Gagal memuat data utang";
      } finally {
        this.loading = false;
      }
    },

    async fetchRingkasan() {
      try {
        const res = await api.get("/api/utang/ringkasan");
        this.ringkasan = res.data;
      } catch (e) {
        // silent
      }
    },

    async tambahUtang(data: {
      nama_pelanggan: string;
      jumlah: number;
      keterangan?: string;
    }) {
      const res = await api.post("/api/utang/", data);
      this.utangList.unshift(res.data);
      await this.fetchRingkasan();
      return res.data;
    },

    async tandaiLunas(id: number) {
      const res = await api.put(`/api/utang/${id}/lunas`);
      const index = this.utangList.findIndex((u) => u.id === id);
      if (index !== -1) {
        this.utangList[index] = res.data;
      }
      await this.fetchRingkasan();
      return res.data;
    },
  },
});
